ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -46,38 +40,59 @@ private[memory] class InMemoryState(
     val state: mutable.Map[ByteString, Array[Byte]] = mutable.Map.empty,
 )
 
+// Dispatcher and LedgerStateAccess are passed in to allow for a multi-participant setup.
 final class InMemoryLedgerReaderWriter(
     override val ledgerId: LedgerId,
     override val participantId: ParticipantId,
     now: () => Instant,
     dispatcher: Dispatcher[Index],
+    ledgerStateAccess: InMemoryLedgerStateAccess,
 )(
     implicit executionContext: ExecutionContext,
     logCtx: LoggingContext,
 ) extends LedgerWriter
     with LedgerReader {
 
-  private val currentState = new InMemoryState()
-
-  private val lockCurrentState = new Semaphore(1, true)
-
   private val committer = new ValidatingCommitter(
     participantId,
     now,
-    SubmissionValidator.create(InMemoryLedgerStateAccess, () => sequentialLogEntryId.next()),
+    SubmissionValidator.create(ledgerStateAccess, () => sequentialLogEntryId.next()),
     dispatcher.signalNewHead,
   )
 
-  private object InMemoryLedgerStateAccess extends LedgerStateAccess[Index] {
-    override def inTransaction[T](body: LedgerStateOperations[Index] => Future[T]): Future[T] =
-      Future
-        .successful(lockCurrentState.acquire())
-        .flatMap(_ => body(InMemoryLedgerStateOperations))
-        .andThen {
-          case _ =>
-            lockCurrentState.release()
-        }
+  override def currentHealth(): HealthStatus =
+    Healthy
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    committer.commit(correlationId, envelope)
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset
+          .map(_.components.head.toInt)
+          .getOrElse(StartIndex),
+        OneAfterAnother[Int, List[LedgerRecord]](
+          (index: Int, _) => index + 1,
+          (index: Int) => Future.successful(List(retrieveLogEntry(index))),
+        ),
+      )
+      .mapConcat { case (_, updates) => updates }
+
+  private def retrieveLogEntry(index: Int): LedgerRecord = {
+    val logEntry = ledgerStateAccess.currentState.log.synchronized {
```

Comment:
If you used the same synchronization primitive as for reading/writing state then you could make `currentState` private and remove `synchronized` blocks when accessing the log. It may make sense to introduce a 'log lock' to gate access to  `currentState.log`.

Reply:
I was thinking the same, but wasn't sure.